### PR TITLE
Dynamic import

### PR DESCRIPTION
**Pull request by piscisaureus**
My attempt at fixing #2510.
The scope went a little overboard.
cc @ry

### CODE REVIEW

**Comment by piscisaureus** [comment]
The core point of the fix is here: https://github.com/denoland/deno/pull/2516/files#diff-7170d073ffb6b76573055a7aaba20d7fR568

File: core/isolate.rs [hunk 1]
```diff
@@ -13,9 +13,13 @@ use crate::libdeno::deno_pinned_buf;
 use crate::libdeno::PinnedBuf;
 use crate::libdeno::Snapshot1;
 use crate::libdeno::Snapshot2;
+use crate::modules::LoadResult;
```
Comment (ry) [inline core/isolate.rs]: isolate.rs shouldn't depend on modules.rs

File: core/isolate.rs [hunk 2]
```diff
@@ -70,29 +74,62 @@ pub enum StartupData<'a> {
...
+}
+
+impl<E> AsImportError for E
+where
+  E: std::error::Error + Sync + 'static,
+{
+  fn as_import_error(self) -> ImportError {
+    Box::new(self)
+  }
+}
```
Comment (ry) [inline core/isolate.rs]: Can a From trait be used instead of AsImportError?

File: core/isolate.rs [hunk 3]
```diff
@@ -109,7 +146,7 @@ pub struct Isolate {
   needs_init: bool,
   shared: SharedQueue,
   pending_ops: FuturesUnordered<OpAsyncFuture>,
-  pending_dyn_imports: FuturesUnordered<DynImport>,
+  pending_dyn_imports: FuturesUnordered<StreamFuture<DynImport>>,
```
Comment (ry) [inline core/isolate.rs]: I don't get why this is a stream...

File: core/modules.rs [hunk 4]
```diff
@@ -140,6 +161,24 @@ impl<L: Loader> RecursiveLoad<L> {
 
     Ok(module_name)
   }
+
+  pub fn get_future(
```
Comment (piscisaureus) [inline core/modules.rs]: Eventually this should go. I added it to make porting over the tests easier.

**Comment by ry** [comment]
ref #1789

File: cli/state.rs [hunk 5]
```diff
@@ -155,6 +155,9 @@ impl Loader for ThreadSafeState {
     referrer: &str,
     is_root: bool,
   ) -> Result<ModuleSpecifier, ErrBox> {
+    // Why are import maps not applicable to the root module?
```
Comment (bartlomieju) [inline cli/state.rs]: This is probably wrong... When I was implementing import maps I believe spec said that it shouldn't be applied to root module and main module of workers, but I can't find this bit now. I'll ask about it and get back to you.

Comment (piscisaureus) [reply cli/state.rs]: I changed it. Now they are still not applied to the main module but they are applied to dynamic imports

Comment (bartlomieju) [reply cli/state.rs]: Reference: https://github.com/WICG/import-maps/issues/157 - it seems that should be changed a bit, but that's a story for separate PR

**Comment by ry** [comment]
Update: This is close to landing and still being worked on.

File: core/isolate.rs [hunk 6]
```diff
@@ -15,9 +15,13 @@ use crate::libdeno::deno_pinned_buf;
 use crate::libdeno::PinnedBuf;
 use crate::libdeno::Snapshot1;
 use crate::libdeno::Snapshot2;
+use crate::modules::RecursiveLoadEvent;
+use crate::modules::SourceCodeInfo;
```
Comment (ry) [inline core/isolate.rs]: This breaks the comment at the top of this file:
```
// Do not add dependenies to modules.rs. it should remain decoupled from the
// isolate to keep the Isolate struct from becoming too bloating for users who
// do not need asynchronous module loading.
```

File: core/modules.rs [hunk 7]
```diff
@@ -886,47 +957,53 @@ mod tests {
...
+        );
+      }
+
+      for _ in 0..10 {
+        let result = recursive_load.poll();
+        assert!(result.is_ok());
+        assert!(result.ok().unwrap().is_not_ready());
+        let l = loads.lock().unwrap();;
+        println!("l {:?}", l);
+        // TODO This is properly broken. FIX BEFORE LANDING.
```
Comment (ry) [inline core/modules.rs]: This should be fixed before landing.

File: core/isolate.rs [hunk 8]
```diff
@@ -895,24 +940,29 @@ pub mod tests {
     })
   }
 
-  #[test]
+  #[cfg(disabled)]
```
Comment (ry) [inline core/isolate.rs]: This should be fixed before landing.

Comment (ry) [reply core/isolate.rs]: Fixed.

File: core/dyn_imports.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+impl Stream for DynImport {
+  type Item = (deno_dyn_import_id, RecursiveLoadEvent);
+  type Error = (deno_dyn_import_id, ErrBox);
+
+  fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+    match self.inner.poll() {
+      Ok(Ready(Some(event))) => Ok(Ready(Some((self.id, event)))),
+      Ok(Ready(None)) => Ok(Ready(None)), // Should never happen.
```
Comment (bartlomieju) [inline core/dyn_imports.rs]: > // Should never happen.

Then maybe `unreachable!()`?

File: core/isolate.rs [hunk 10]
```diff
@@ -57,6 +65,21 @@ pub struct Script<'a> {
...
 
+/// Represent result of fetching the source code of a module.
+/// Contains both module name and code.
+/// Module name might be different from initial URL used for loading
+/// due to redirections.
+/// e.g. Both https://example.com/a.ts and https://example.com/b.ts
+/// may point to https://example.com/c.ts. By specifying module_name
+/// all be https://example.com/c.ts in module_name (for aliasing),
+/// we avoid recompiling the same code for 3 different times.
+#[derive(Debug, Eq, PartialEq)]
```
Comment (bartlomieju) [inline core/isolate.rs]:
```
/// Represent result of fetching the source code of a module.
/// Found module URL might be different from specified URL used for loading
/// due to redirections.
/// e.g. Both https://example.com/a.ts and https://example.com/b.ts
/// may point to https://example.com/c.ts. 
/// By keeping track of specified and found URL we can alias
/// modules and avoid recompiling the same code 3 times.
```

File: core/modules.rs [hunk 11]
```diff
@@ -55,209 +49,269 @@ pub trait Loader: Send + Sync {
...
+      let root_id = match self.state {
+        State::LoadingImports(mod_id) => mod_id,
+        _ => unreachable!(),
+      };
+
+      let mut resolve_cb =
+        |specifier: &str, referrer_id: deno_mod| -> deno_mod {
+          let modules = self.modules.lock().unwrap();
+          let referrer = modules.get_name(referrer_id).unwrap();
+          match self.loader.resolve(specifier, &referrer, false) {
```
Comment (bartlomieju) [inline core/modules.rs]: Question: is `resolve_cb` called only for root module? I believe during writing import maps I asserted it is root module, but I'm not sure now. If it's not a root then hardcoded `false` should be changed

File: core/modules.rs [hunk 12]
```diff
@@ -55,209 +49,269 @@ pub trait Loader: Send + Sync {
...
+  LoadingRoot,
+  LoadingImports(deno_mod),
+  Instantiated(deno_mod),
 }
 
 /// This future is used to implement parallel async module loading without
-/// complicating the Isolate API. Note that RecursiveLoad will take ownership of
-/// an Isolate during load.
+/// complicating the Isolate API.
+/// TODO: RecursiveLoad desperately needs to be merged with Modules.
```
Comment (bartlomieju) [inline core/modules.rs]: Agreed, I understood what was going on in `modules.rs`... not anymore

File: core/modules.rs [hunk 13]
```diff
@@ -55,209 +49,269 @@ pub trait Loader: Send + Sync {
...
+    // 2. Source code resolved result have a different name as requested:
+    //   2a. The module with resolved module name has been registered
+    //     -> alias
+    //   2b. The module with resolved module name has not yet been registerd
+    //     -> register & alias
+    let SourceCodeInfo {
+      code,
+      module_url_specified,
+      module_url_found,
+    } = source_code_info;
```
Comment (bartlomieju) [inline core/modules.rs]: Didn't know you could do this in Rust, nice :+1:

File: core/modules.rs (hunk 11 above)
Comment (ry) [reply core/modules.rs]: correct. (but now the root module is no longer the same as the main module - it's just the root of any RecursiveLoad)

File: core/isolate.rs (hunk 10 above)
Comment (ry) [reply core/isolate.rs]: Thanks - updated.

File: core/modules.rs (hunk 11 above)
Comment (bartlomieju) [reply core/modules.rs]: > (but now the root module is no longer the same as the main module - it's just the root of any RecursiveLoad)

So if I understand correctly then we should change that `false` to actual comparison if this module is main module

File: core/dyn_imports.rs (hunk 9 above)
Comment (ry) [reply core/dyn_imports.rs]: Done

File: core/modules.rs (hunk 11 above)
Comment (ry) [reply core/modules.rs]: Done. We'll need to have some tests soon showing import maps working against dynamic import - this change had no effect on the tests.

File: core/isolate.rs (hunk 1 above)
Comment (ry) [reply core/isolate.rs]: fixed.

File: core/isolate.rs (hunk 6 above)
Comment (ry) [reply core/isolate.rs]: fixed.

File: core/modules.rs (hunk 7 above)
Comment (ry) [reply core/modules.rs]: Fixed

**Comment by ry** [comment]
On appveyor the unit tests are failing. No idea why.
```
Summary for <no permissions>
undefined
Summary for --allow-read
test result: OK. 34 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (58.00ms)
Summary for --allow-net
test result: OK. 23 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (5514.00ms)
Summary for --allow-read --allow-write
test result: OK. 47 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (80.00ms)
Summary for --allow-write
test result: OK. 16 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (20.00ms)
Summary for --allow-env
test result: OK. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (10.00ms)
Summary for --allow-run
test result: OK. 12 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (254.00ms)
Summary for --allow-write --allow-run
test result: OK. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (36.00ms)
Summary for --allow-read --allow-write --allow-run
test result: OK. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (72.00ms)
Summary for --allow-hrtime
test result: OK. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (-0.16ms)
Unit tests failed
FAIL
```

**Comment by piscisaureus** [comment]
https://github.com/denoland/deno/pull/2516/files#diff-e42bc8cc82b66eecdca76f5c6069e716?

**Comment by ry** [comment]
It's disabled https://github.com/denoland/deno/pull/2516/files#diff-06291984cb0c2f3e0fad6822ab277aec and not part of the unit tests.

I think we can land without error_014_catch_dynamic_import_error.js working yet.

**Pull request merged**